### ISSUE TITLE
Play the environment with other RL algorithms

### ISSUE THREAD

**Issue by Screamer-Y** [microsoft/CyberBattleSim#91]
Hi everyone,
I'm interested in how the environment in this project performs under other reinforcement algorithms. But I'm new to reinforcement learning and not yet capable of implementing other reinforcement learning algorithms on my own. 
I noticed that algorithms such as Q-Learning and DQN have been implemented in the baselines directory. And the definition of the environment is quite different from some basic gym environments (e.g. CartPole), which made it difficult for me to understand.
I wonder if it is possible to implement different RL algorithms in the environments of this project with the help of some RL algorithm libraries? (e.g. Stable-Baselines3, Keras-rl)
I would greatly appreciate the help.
Thank you

**Comment by 2twentytwo2** [comment]
您好，您的邮件我已收到，
看到后，我会立即回复您。
This is an automatic reply, confirming that your e-mail was received.Thank you

**Comment by Screamer-Y** [comment]
Hi everyone,
I noticed that PR(#86) already has support for stable-baselines3 and gives the corresponding use cases in the notebooks folder, which is of greate help. Sorry for missing that before...

**Comment by kvas7andy** [comment]
Hi, @Screamer-Y, actually that is a interesting topic for me to investigate. If you will have any problems or you find any literature using this simulator, feel free to share, and I can do the same if you are interested.

I am testing stable-baselines now and have errors with [https://github.com/microsoft/CyberBattleSim/blob/4fd228bccfc2b088d911e27072a923251203cac8/notebooks/stable-baselines-agent.py] (stable-baselines-agent.py)

`  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/root/.vscode-server/extensions/ms-python.python-2022.18.2/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/__main__.py", line 39, in <module>
    cli.main()
  File "/root/.vscode-server/extensions/ms-python.python-2022.18.2/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 430, in main
    run()
  File "/root/.vscode-server/extensions/ms-python.python-2022.18.2/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher/../../debugpy/../debugpy/server/cli.py", line 284, in run_file
    runpy.run_path(target, run_name="__main__")
  File "/root/.vscode-server/extensions/ms-python.python-2022.18.2/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 321, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/root/.vscode-server/extensions/ms-python.python-2022.18.2/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 135, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/root/.vscode-server/extensions/ms-python.python-2022.18.2/pythonFiles/lib/python/debugpy/_vendored/pydevd/_pydevd_bundle/pydevd_runpy.py", line 124, in _run_code
    exec(code, run_globals)
  File "/root/notebooks/stable-baselines-agent.py", line 54, in <module>
    model_a2c = A2C("MultiInputPolicy", env2).learn(10000)
  File "/usr/local/lib/python3.8/dist-packages/stable_baselines3/a2c/a2c.py", line 203, in learn
    return super().learn(
  File "/usr/local/lib/python3.8/dist-packages/stable_baselines3/common/on_policy_algorithm.py", line 262, in learn
    continue_training = self.collect_rollouts(self.env, callback, self.rollout_buffer, n_rollout_steps=self.n_steps)
  File "/usr/local/lib/python3.8/dist-packages/stable_baselines3/common/on_policy_algorithm.py", line 172, in collect_rollouts
    actions, values, log_probs = self.policy(obs_tensor)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/stable_baselines3/common/policies.py", line 586, in forward
    features = self.extract_features(obs)
  File "/usr/local/lib/python3.8/dist-packages/stable_baselines3/common/policies.py", line 126, in extract_features
    preprocessed_obs = preprocess_obs(obs, self.observation_space, normalize_images=self.normalize_images)
  File "/usr/local/lib/python3.8/dist-packages/stable_baselines3/common/preprocessing.py", line 127, in preprocess_obs
    preprocessed_obs[key] = preprocess_obs(_obs, observation_space[key], normalize_images=normalize_images)
  File "/usr/local/lib/python3.8/dist-packages/stable_baselines3/common/preprocessing.py", line 108, in preprocess_obs
    return F.one_hot(obs.long(), num_classes=observation_space.n).float()
RuntimeError: num_classes should be positive`

**Comment by Screamer-Y** [comment]
> Hi, @Screamer-Y, actually that is a interesting topic for me to investigate. If you will have any problems or you find any literature using this simulator, feel free to share, and I can do the same if you are interested.
> 
> Edited: I found the source of the error it was purely because of my own additions to the simulator.

Hi, @kvas7andy , it's glad to know you're interested in this topic too. 
Now I'm working on [https://github.com/microsoft/CyberBattleSim/blob/4fd228bccfc2b088d911e27072a923251203cac8/cyberbattle/_env/flatten_wrapper.py](url).
My goal is to modify the 'action_space' from a 'spaces.MultiDiscrete' to a 'spaces.Discrete'. I do this because in stable-baselines3, if you want to use the value-based RL algorithm, action_space can only be 'spaces.Discrete'. Currently I have simply mapped all the possible actions to a Discrete value, but when a try to train a DQN agent, it can not learn from the environment properly. I'm still trying to figure out what went wrong.
Here is my modification,
```python
class FlattenActionWrapper(ActionWrapper):
    """
    Flatten all nested dictionaries and tuples from the
     action space of a CyberBattleSim environment`CyberBattleEnv`.
     The resulting action space is a `Discrete`.
    """

    def __init__(self, env: CyberBattleEnv):
        ActionWrapper.__init__(self, env)
        self.env = env

        # local:source_node_num x local_attacks_count;remote:source x target x remote_attacks_count;
        # connect:source x target x remote_attacks_count x port x credentials
        self.action_space = spaces.Discrete(env.bounds.maximum_node_count*env.bounds.local_attacks_count + \
            env.bounds.maximum_node_count*env.bounds.maximum_node_count*env.bounds.remote_attacks_count + \
            env.bounds.maximum_node_count*env.bounds.maximum_node_count*env.bounds.port_count*env.bounds.maximum_total_credentials)
        

    def action(self, action: np.int64) -> Action:
        n_nodes = self.env.bounds.maximum_node_count
        n_local_attacks = self.env.bounds.local_attacks_count
        n_remote_attacks = self.env.bounds.remote_attacks_count
        n_port = self.env.bounds.port_count
        n_credentials = self.env.bounds.maximum_total_credentials

        if action<n_nodes*n_local_attacks:
            source_node = action//n_local_attacks
            local_vulnerability = action%n_local_attacks
            return {'local_vulnerability': np.array([source_node, local_vulnerability])}

        action -= n_nodes*n_local_attacks
        if action < n_nodes*n_nodes*n_remote_attacks:
            source_node = action//(n_remote_attacks*n_nodes)
            target_node = (action//n_remote_attacks)%n_nodes
            remote_vulnerability = action%n_remote_attacks
            return {'remote_vulnerability': np.array([source_node, target_node, remote_vulnerability])}

        action -= n_nodes*n_nodes*n_remote_attacks
        if action < n_nodes*n_nodes*n_port*n_credentials:
            source_node = action//(n_nodes*n_port*n_credentials)
            target_node = (action//(n_port*n_credentials))%n_nodes
            port = (action//(n_credentials))%n_port
            credential = action%n_credentials
            return {'connect': np.array([source_node,target_node,port,credential])}

        raise NotSupportedError(f'Unsupported action: {action}')

    def reverse_action(self, action):
        raise NotImplementedError
```
I'm not a good programmer, so feel free to point out any problem and I will appreciate it.

**Comment by forrestmckee** [comment]
> > Hi, @Screamer-Y, actually that is a interesting topic for me to investigate. If you will have any problems or you find any literature using this simulator, feel free to share, and I can do the same if you are interested.
> > Edited: I found the source of the error it was purely because of my own additions to the simulator.
> 
> Hi, @kvas7andy , it's glad to know you're interested in this topic too. Now I'm working on [https://github.com/microsoft/CyberBattleSim/blob/4fd228bccfc2b088d911e27072a923251203cac8/cyberbattle/_env/flatten_wrapper.py](url). My goal is to modify the 'action_space' from a 'spaces.MultiDiscrete' to a 'spaces.Discrete'. I do this because in stable-baselines3, if you want to use the value-based RL algorithm, action_space can only be 'spaces.Discrete'. Currently I have simply mapped all the possible actions to a Discrete value, but when a try to train a DQN agent, it can not learn from the environment properly. I'm still trying to figure out what went wrong. Here is my modification,
> 
> ```python
> class FlattenActionWrapper(ActionWrapper):
>     """
>     Flatten all nested dictionaries and tuples from the
>      action space of a CyberBattleSim environment`CyberBattleEnv`.
>      The resulting action space is a `Discrete`.
>     """
> 
>     def __init__(self, env: CyberBattleEnv):
>         ActionWrapper.__init__(self, env)
>         self.env = env
> 
>         # local:source_node_num x local_attacks_count;remote:source x target x remote_attacks_count;
>         # connect:source x target x remote_attacks_count x port x credentials
>         self.action_space = spaces.Discrete(env.bounds.maximum_node_count*env.bounds.local_attacks_count + \
>             env.bounds.maximum_node_count*env.bounds.maximum_node_count*env.bounds.remote_attacks_count + \
>             env.bounds.maximum_node_count*env.bounds.maximum_node_count*env.bounds.port_count*env.bounds.maximum_total_credentials)
>         
> 
>     def action(self, action: np.int64) -> Action:
>         n_nodes = self.env.bounds.maximum_node_count
>         n_local_attacks = self.env.bounds.local_attacks_count
>         n_remote_attacks = self.env.bounds.remote_attacks_count
>         n_port = self.env.bounds.port_count
>         n_credentials = self.env.bounds.maximum_total_credentials
> 
>         if action<n_nodes*n_local_attacks:
>             source_node = action//n_local_attacks
>             local_vulnerability = action%n_local_attacks
>             return {'local_vulnerability': np.array([source_node, local_vulnerability])}
> 
>         action -= n_nodes*n_local_attacks
>         if action < n_nodes*n_nodes*n_remote_attacks:
>             source_node = action//(n_remote_attacks*n_nodes)
>             target_node = (action//n_remote_attacks)%n_nodes
>             remote_vulnerability = action%n_remote_attacks
>             return {'remote_vulnerability': np.array([source_node, target_node, remote_vulnerability])}
> 
>         action -= n_nodes*n_nodes*n_remote_attacks
>         if action < n_nodes*n_nodes*n_port*n_credentials:
>             source_node = action//(n_nodes*n_port*n_credentials)
>             target_node = (action//(n_port*n_credentials))%n_nodes
>             port = (action//(n_credentials))%n_port
>             credential = action%n_credentials
>             return {'connect': np.array([source_node,target_node,port,credential])}
> 
>         raise NotSupportedError(f'Unsupported action: {action}')
> 
>     def reverse_action(self, action):
>         raise NotImplementedError
> ```
> 
> I'm not a good programmer, so feel free to point out any problem and I will appreciate it.

@Screamer-Y, did you get the stable-baselines example script to work? For me it runs, but the agent never learns anything using A2C or PPO.

**Comment by Screamer-Y** [comment]
> @Screamer-Y, did you get the stable-baselines example script to work? For me it runs, but the agent never learns anything using A2C or PPO.
Hi @forrestmckee ,
Yes, From my side it works properly, I just run the code in [https://github.com/microsoft/CyberBattleSim/blob/main/notebooks/stable-baselines-agent.py](url) without any modification.

**Comment by forrestmckee** [comment]
@Screamer-Y are you using Linux, WSL, or Docker?

I can get the script you referenced to run, but the agent never makes it off of the foothold node regardless of the number of time steps I set. I'm also getting warnings that the agent is trying to access an invalid index.

**Comment by Screamer-Y** [comment]
> @Screamer-Y are you using Linux, WSL, or Docker?
> 
> I can get the script you referenced to run, but the agent never makes it off of the foothold node regardless of the number of time steps I set. I'm also getting warnings that the agent is trying to access an invalid index.

Hi @forrestmckee ,
I'm using Ubuntu Server 20.04 LTS. I ran the script again just now and only made one successful connect action with 10000 time steps. I think the problem is due to the way 'action_space' defined in ['flatten_wrapper'](https://github.com/microsoft/CyberBattleSim/blob/4fd228bccfc2b088d911e27072a923251203cac8/cyberbattle/_env/flatten_wrapper.py), which contains all attacks, even if it is invalid and it's also the reason why you keep getting warnings.
I have the same problem when turning the 'action_space' into 'spaces.Discrete', one possible solution is reduce the dims of 'action_space' just as the way in ['agent_wrapper].(https://github.com/microsoft/CyberBattleSim/blob/main/cyberbattle/agents/baseline/agent_wrapper.py)

**Comment by Gabriel0402** [comment]
@forrestmckee I came across the same issue as you met. But I noticed another interesting thing: during the training, although we got the warnings that the agent is trying to access an invalid index, but the number of nodes discovered so far is increasing. I think this means that the A2C or PPO is actually working. They did discover new nodes. The thing I don't understand is when the trained model is applied to the action prediction, it never discovers new nodes. 

@Screamer-Y I don't quite understand why warnings are incorrect. I think you will also see the warnings if you set the logging levels. Because we have to discover new nodes, whose number is less than the maximum node count. So when the nodes are not discovered, we will always get warnings. And I also don't understand why we have to reduce the dims of action space.

**Comment by forrestmckee** [comment]
@blumu Is there a planned `sample_valid_action` equivalent for Flattened Environments/Stable Baselines3? I believe what myself and others have discovered is that the entire observation and action spaces are "fair game" for the agent to sample from at any given time. Doesn't this mean that an agent can attempt to take an action both to and  from a node that it hasn't discovered yet? This seems to greatly increase the number of time steps required for an agent to learn. 

@Screamer-Y were you able to reduce the dims of the action space like you mentioned?

**Comment by blumu** [comment]
@forrestmckee  That's a good suggestion, it's not currently planned but it feels like something we ought to have. Work on stable_baselines3 is currently postponed until they complete the upgrade to the latest version of gym. Once this is done the plan is to upgrade cyberbattlesim to the latest version of gym and stable_baselines which will then allow for further improvements like the one you mentioned.

**Comment by Screamer-Y** [comment]
Thanks for all the suggestions!
@Gabriel0402 I think you are right about the warnings, I didn't have a good understanding of the code at the time.
 
Regarding the second question, I once expected to speed up the learning process by reducing the size of action_space, and after trying this I found that this did not work significantly.

@forrestmckee @blumu So I still have questions: Is there a significant performance difference between the A2C or PPO methods implemented in 
[Stable-baseline3](https://github.com/microsoft/CyberBattleSim/blob/main/notebooks/stable-baselines-agent.py ) and the DQN method implemented in [agent_dql](https://github.com/microsoft/CyberBattleSim/blob/main/cyberbattle/agents/baseline/agent_dql.py)?

With the same set of iteration_count=1500 and episode_count=20, I observed in toy-ctf that A2C only gets an average return of no more than 40 per episode, which is far from the average return of about 450 in [benchmark](https://github.com/microsoft/CyberBattleSim/blob/main/notebooks/notebook_benchmark-toyctf.ipynb). I would be very grateful if you have any better approaches to improve the performance of Stable-baseline3.

**Comment by forrestmckee** [comment]
@Screamer-Y the difference between the built in algorithms and Stable-Baselines3 is that the built ins have a check to ensure a valid action. SB3 doesn't, so a large portion of the time you're performing an impossible action given the current state of the environment.

**Comment by Screamer-Y** [comment]
@forrestmckee Thank you so much for the speedy reply. I think I've understood what you've mentioned in this comment and the previous one :)

**Comment by abdullahiisaahmed** [comment]
> Hi, [@Screamer-Y](https://github.com/Screamer-Y), actually that is a interesting topic for me to investigate. If you will have any problems or you find any literature using this simulator, feel free to share, and I can do the same if you are interested.
> 
> Edited: I found the source of the error it was purely because of my own additions to the simulator.

I just found a paper that uses this simulator. Kindly share if you found any from your end. Thanks

**Paper1**: [Enhancing Microsoft CyberBattleSim for Enterprise Cybersecurity Simulations](https://www.scirp.org/journal/paperinformation?paperid=141811)

**Paper2**: [A Multiagent CyberBattleSim for RL Cyber Operation Agents](https://arxiv.org/pdf/2304.11052)
